Faces: Improve performance and strategy for manual tagging

theshadow27 opened the pull request:
# (1) Please provide a concise description of your pull request
## What does it implement / fix / improve? Why?

**What?**: Per [#3124](https://github.com/photoprism/photoprism/issues/3124), this PR implements strategy [(1)](https://github.com/photoprism/photoprism/issues/3124#issuecomment-2558299360) to improve `Optimize()` performance by returning the correct number of rows modified in `PurgeOrphanFaces` and  treating the result `0` as an error condition in `MergeFaces`. Logging around this has been improved in `faces.go` and `faces_optimize.go`. 

**Why?**: To improve performance when manually tagging faces, in the case that it becomes necessary to call `Optimize()`. Please see [issue 3124](https://github.com/photoprism/photoprism/issues/3124#issuecomment-2558299360) for the detailed description of the problem. 

## Are the changes related to an existing issue?

Yes: [#3124](https://github.com/photoprism/photoprism/issues/3124)

# (2) CLA (see https://www.photoprism.app/cla)

Signed and scanned

# (3) Acceptance Criteria:

- [x] Features and enhancements must be fully implemented so that they can be released at any time without additional work
- [ ] Automated unit and/or acceptance tests are mandatory to ensure the changes work as expected and to reduce repetitive manual work *see below!*
- [x] Frontend components must be responsive to work and look properly on phones, tablets, and desktop computers; you must have tested them on all major browsers and different devices *- no frontend changes, but did test on chrome, firefox, safari*
- [x] Documentation and translation updates should be provided if needed *no translations modified*
- [x] In case you submit database-related changes, they must be tested and compatible with SQLite 3 and MariaDB 10.5.12+ *no database schema changes*

<!--

Since reviewing, testing and finally merging pull requests requires significant resources on our side, this can take several months if it's not just a small fix, especially if extensive testing is required to prevent bugs from getting into our stable version.

We thank you for your patience! :)

-->
# Testing issues 

Development machine is Windows 10 (Docker WSLv2). Followed instructions, having test errors that seem disconnected from the modified code:

```
FAIL | UpdateMarker (0.08s)
FAIL |   UpdateMarker/update_cluster_with_existing_subject_2 (0.02s)
     |         markers_test.go:272:
     |                  Error Trace:    /go/src/github.com/photoprism/photoprism/internal/api/markers_test.go:272
     |                  Error:          Not equal:
     |                                  expected: 200
     |                                  actual  : 500
     |                  Test:           TestUpdateMarker/update_cluster_with_existing_subject_2
...
     | time="2024-12-30T18:39:51Z" level=debug msg="api-v1: abort /api/v1/zip with code 400 (unable to do that)"
     | time="2024-12-30T18:39:51Z" level=error msg="download: /go/src/github.com/photoprism/photoprism/storage/testdata/temp/zip/xxx not found"
FAIL |  github.com/photoprism/photoprism/internal/api   72.942s
...
FAIL | MediaFile_NeedsExifToolJson (0.12s)
FAIL |   MediaFile_NeedsExifToolJson/false (0.04s)
     |         mediafile_meta_test.go:84:
     |                  Error Trace:    /go/src/github.com/photoprism/photoprism/internal/photoprism/mediafile_meta_test.go:84
     |                  Error:          Should be true
     |                  Test:           TestMediaFile_NeedsExifToolJson/false
FAIL |  github.com/photoprism/photoprism/internal/photoprism
```

FWIW, in my environment, these tests **all fail on a completely vanilla develop branch**. During my attempt at implementation, there were other errors/failing tests that *were* caught and are now passing, so I assume that these test failures are alright. 

In lieu of passing tests, I built using `make preview` (changing the docker-buildx.sh script to use `--load` instead of `--push`) and set the image in my main photoprism deployment (after backing everything up) to use the local image. Observations will follow in a comment below.

theshadow27 committed:
set err when no faces merged

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/photoprism/photoprism?pullRequest=4691) <br/>All committers have signed the CLA.

theshadow27 commented:
Unfortunately, this first try did not work. Still seeing the same behavior, but at least it's complaining about it now:

```
2024-12-30 16:51:27 time="2024-12-30T21:51:27Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:51:27 time="2024-12-30T21:51:27Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:51:27 time="2024-12-30T21:51:27Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:51:27 time="2024-12-30T21:51:27Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 0 cluster 123 count 5"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.005618"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.140689"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with ZA7UNK35CGLE6NVJWZ3JJRUXMDSRVWBP, subject 'RedactedName0', dist 0.000936"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.006674"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.124564"
2024-12-30 16:51:38 time="2024-12-30T21:51:38Z" level=debug msg="faces: can merge 536V226VWHHQJRZDYXMRUQVNQCBMXXX3 with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.008516"
2024-12-30 16:51:47 time="2024-12-30T21:51:47Z" level=debug msg="faces: removed 2 orphans of 7 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:51:48 time="2024-12-30T21:51:48Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:51:48 time="2024-12-30T21:51:48Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:51:48 time="2024-12-30T21:51:48Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:51:48 time="2024-12-30T21:51:48Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:51:57 time="2024-12-30T21:51:57Z" level=debug msg="server: GET /api/v1/config (200) [1.421822ms]"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 1 cluster 123 count 5"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: can merge 5ZB5AA2VX4FUSA4LNXGU5RPPOM3TMVNF with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004816"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: can merge 5ZB5AA2VX4FUSA4LNXGU5RPPOM3TMVNF with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141492"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: can merge 5ZB5AA2VX4FUSA4LNXGU5RPPOM3TMVNF with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005871"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: can merge 5ZB5AA2VX4FUSA4LNXGU5RPPOM3TMVNF with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123761"
2024-12-30 16:51:58 time="2024-12-30T21:51:58Z" level=debug msg="faces: can merge 5ZB5AA2VX4FUSA4LNXGU5RPPOM3TMVNF with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007713"
2024-12-30 16:52:07 time="2024-12-30T21:52:07Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:52:07 time="2024-12-30T21:52:07Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:52:07 time="2024-12-30T21:52:07Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:52:07 time="2024-12-30T21:52:07Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:52:07 time="2024-12-30T21:52:07Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:52:17 time="2024-12-30T21:52:17Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:52:17 time="2024-12-30T21:52:17Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 2 cluster 123 count 5"
2024-12-30 16:52:18 time="2024-12-30T21:52:18Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with FYBFU3P3HTI2JOUBS6LSYJUMFHCSQOVI, subject 'RedactedName0', dist 0.004704"
2024-12-30 16:52:18 time="2024-12-30T21:52:18Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.146307"
2024-12-30 16:52:18 time="2024-12-30T21:52:18Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.001055"
2024-12-30 16:52:18 time="2024-12-30T21:52:18Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.118946"
2024-12-30 16:52:18 time="2024-12-30T21:52:18Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.002897"
2024-12-30 16:52:26 time="2024-12-30T21:52:26Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:52:26 time="2024-12-30T21:52:26Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:52:26 time="2024-12-30T21:52:26Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:52:26 time="2024-12-30T21:52:26Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:52:26 time="2024-12-30T21:52:26Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 3 cluster 123 count 5"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: can merge 4GZUGRU3E4VBA4ROADTHOG6JC4TKMITD with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004686"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: can merge 4GZUGRU3E4VBA4ROADTHOG6JC4TKMITD with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141622"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: can merge 4GZUGRU3E4VBA4ROADTHOG6JC4TKMITD with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005741"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: can merge 4GZUGRU3E4VBA4ROADTHOG6JC4TKMITD with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123631"
2024-12-30 16:52:37 time="2024-12-30T21:52:37Z" level=debug msg="faces: can merge 4GZUGRU3E4VBA4ROADTHOG6JC4TKMITD with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007583"
2024-12-30 16:52:45 time="2024-12-30T21:52:45Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:52:45 time="2024-12-30T21:52:45Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:52:45 time="2024-12-30T21:52:45Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:52:45 time="2024-12-30T21:52:45Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:52:45 time="2024-12-30T21:52:45Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 4 cluster 123 count 5"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.146307"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with VFW56NHKUZ37WCJYLV62NV62UZAWG4J4, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.001055"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.118946"
2024-12-30 16:52:56 time="2024-12-30T21:52:56Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.002897"
2024-12-30 16:53:05 time="2024-12-30T21:53:05Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:53:05 time="2024-12-30T21:53:05Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:53:05 time="2024-12-30T21:53:05Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:53:05 time="2024-12-30T21:53:05Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:53:05 time="2024-12-30T21:53:05Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 5 cluster 123 count 5"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: can merge 7LZ27HJU6AHYXUBUSWK3H7QUWMB5G2GL with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: can merge 7LZ27HJU6AHYXUBUSWK3H7QUWMB5G2GL with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141625"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: can merge 7LZ27HJU6AHYXUBUSWK3H7QUWMB5G2GL with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005737"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: can merge 7LZ27HJU6AHYXUBUSWK3H7QUWMB5G2GL with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123627"
2024-12-30 16:53:15 time="2024-12-30T21:53:15Z" level=debug msg="faces: can merge 7LZ27HJU6AHYXUBUSWK3H7QUWMB5G2GL with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007579"
2024-12-30 16:53:24 time="2024-12-30T21:53:24Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:53:25 time="2024-12-30T21:53:25Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:53:25 time="2024-12-30T21:53:25Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:53:25 time="2024-12-30T21:53:25Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:53:25 time="2024-12-30T21:53:25Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 6 cluster 123 count 5"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: can merge BQBIMWQ5Q6U2VLOTZKTPRGKA3PFVJTBW with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: can merge BQBIMWQ5Q6U2VLOTZKTPRGKA3PFVJTBW with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141626"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: can merge BQBIMWQ5Q6U2VLOTZKTPRGKA3PFVJTBW with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005737"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: can merge BQBIMWQ5Q6U2VLOTZKTPRGKA3PFVJTBW with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123627"
2024-12-30 16:53:35 time="2024-12-30T21:53:35Z" level=debug msg="faces: can merge BQBIMWQ5Q6U2VLOTZKTPRGKA3PFVJTBW with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007579"
2024-12-30 16:53:44 time="2024-12-30T21:53:44Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:53:44 time="2024-12-30T21:53:44Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:53:44 time="2024-12-30T21:53:44Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:53:44 time="2024-12-30T21:53:44Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:53:44 time="2024-12-30T21:53:44Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 7 cluster 123 count 5"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: can merge 7OWPZYIB4KAFL2KHFIEITLP2OVHOQ4OR with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: can merge 7OWPZYIB4KAFL2KHFIEITLP2OVHOQ4OR with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141626"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: can merge 7OWPZYIB4KAFL2KHFIEITLP2OVHOQ4OR with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005737"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: can merge 7OWPZYIB4KAFL2KHFIEITLP2OVHOQ4OR with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123627"
2024-12-30 16:53:54 time="2024-12-30T21:53:54Z" level=debug msg="faces: can merge 7OWPZYIB4KAFL2KHFIEITLP2OVHOQ4OR with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007579"
2024-12-30 16:54:03 time="2024-12-30T21:54:03Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:54:03 time="2024-12-30T21:54:03Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:54:03 time="2024-12-30T21:54:03Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:54:03 time="2024-12-30T21:54:03Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:54:03 time="2024-12-30T21:54:03Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 8 cluster 123 count 5"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: can merge ANOXXTX33HNTUDSW75WJ5IVE6KTHA4L3 with CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: can merge ANOXXTX33HNTUDSW75WJ5IVE6KTHA4L3 with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.141626"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: can merge ANOXXTX33HNTUDSW75WJ5IVE6KTHA4L3 with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.005737"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: can merge ANOXXTX33HNTUDSW75WJ5IVE6KTHA4L3 with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.123627"
2024-12-30 16:54:13 time="2024-12-30T21:54:13Z" level=debug msg="faces: can merge ANOXXTX33HNTUDSW75WJ5IVE6KTHA4L3 with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.007579"
2024-12-30 16:54:22 time="2024-12-30T21:54:22Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:54:22 time="2024-12-30T21:54:22Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:54:22 time="2024-12-30T21:54:22Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:54:22 time="2024-12-30T21:54:22Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:54:22 time="2024-12-30T21:54:22Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 9 cluster 123 count 5"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with FN3BUBP2MPYTJLGKAY3SF5K3NLYHKK6D, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.146307"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.001055"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.118946"
2024-12-30 16:54:32 time="2024-12-30T21:54:32Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.002897"
2024-12-30 16:54:41 time="2024-12-30T21:54:41Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:54:41 time="2024-12-30T21:54:41Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with QMCUGOBSQCPJRRVTNPN4BM4H37BOFJPO, subject 'RedactedName1', dist 0.007817"
2024-12-30 16:54:41 time="2024-12-30T21:54:41Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with 2KGYTKQYWY32QZS6UD2V4A2RN73AABWO, subject 'RedactedName1', dist 0.166765"
2024-12-30 16:54:41 time="2024-12-30T21:54:41Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with DTNACFUL7BVCKLEIKBWKTLOYDSH64LZL, subject 'RedactedName1', dist 0.002085"
2024-12-30 16:54:41 time="2024-12-30T21:54:41Z" level=debug msg="faces: can merge 76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ with JLV7E7A2FU2ED33XHNMUXVT3YI5FY7IL, subject 'RedactedName1', dist 0.133412"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 10 cluster 123 count 5"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with PHDH4QECCHZD6CNHGQIPOWBT4LHAEVA4, subject 'RedactedName0', dist 0.004682"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TD5AHC2UPI7FOMRPQLIZJZFUINQPGYNS, subject 'RedactedName0', dist 0.146307"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with GXOB2FPBHZPVYSKI24HFTNWTZKUH7LE3, subject 'RedactedName0', dist 0.001055"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with HZTEPGU54EKGIU67WEXV3GYGTN6562EZ, subject 'RedactedName0', dist 0.118946"
2024-12-30 16:54:52 time="2024-12-30T21:54:52Z" level=debug msg="faces: can merge CKC6GOANRLSGT6ISEUQEYQMHG4NJHX7Q with TEBA7O57VVBV664QKYJ555FOUWM2Y7JU, subject 'RedactedName0', dist 0.002897"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="faces: removed 1 orphans of 6 canidates for subject jrwocac3n3w65ivq"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=info msg="faces: merged 67 clusters"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="covers: updated 1 subject [230.64982ms]"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="counts: updated 1 subject [332.684788ms]"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="search: updated index by photo id [2.575121ms]"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="photos: found 1 result for uid:psoq2cv5hsbqqw2i dist:2.000000 merged:true [2.927452ms]"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=info msg="changes successfully saved"
2024-12-30 16:55:00 time="2024-12-30T21:55:00Z" level=debug msg="server: PUT /api/v1/markers/msoq2cvgfd0papfd (200) [3m47.071525485s]"
```

Results of preliminary analysis:

The 4 repeating lines regarding merging `76NM6PHBWW6TWWL6KUAH6Q44XVHQBCMZ` as well as 

```
2024-12-30 16:52:17 time="2024-12-30T21:52:17Z" level=debug msg="faces: no affected rows for purge in batch 0 - 5"
2024-12-30 16:52:17 time="2024-12-30T21:52:17Z" level=error msg="faces: failed to remove any orphan clusters of 5 canidates for subject jrr2slz1aipti9h5 (merge) itr 2 cluster 123 count 5"
```

Are expected, since every cluster is tried every time in each iteration. 

The other lines are interesting... This particular subject there are 6 manually tagged clusters and one automatically tagged cluster (after the operations copied above are complete)

![image](https://github.com/user-attachments/assets/cde489cd-5aa0-48fb-a015-290181f65aac)

So:
* Cluster `CKC6...` has 6 samples and is `faces[1]` 6 times, and `faces[0]` **4 times** on itr 2, 4, 9, and 10
* Cluster `TD5A...` has also 6 samples and is always `faces[2]` except for 1 time when it is `faces[1]`
* Cluster `GXOB...` has 5 samples, is `faces[4]` in itr 0 and then `faces[3]` after that.  
* Cluster `HZTE...` has 4 samples, is `faces[5]` in itr 0 and then `faces[4]` after that.
* Cluster `TEBA...` has 4 samples, is faces[6]` in itr 0 and then `faces[5]` after that.

The rest are seen only once:
* Cluster `536V...` is `faces[0]` once on itr 0. Removed 2 of 7
* Cluster `5ZB5...` is `faces[0]` once on itr 1. Removed 1 of 6
* Cluster `4GZU...` is `faces[0]` once on itr 3. Removed 1 of 6
* Cluster `7LZ2...` is `faces[0]` once on itr 5. Removed 1 of 6
* Cluster `BQBI...` is `faces[0]` once on itr 6. Removed 1 of 6
* Cluster `7OWP...` is `faces[0]` once on itr 7. Removed 1 of 6
* Cluster `ANOX...` is `faces[0]` once on itr 8. Removed 1 of 6

From the DBeaver screenshot, there is one unaccounted for cluster, `MH4XYDJ3AEZ3GF32AWJL2OYK6DMPTPP5` that is probably the last one created by this loop. Each time, a new one is created with the centroid of the previous one, plus the centroids of the other manual tags. Somehow, this results in a new face hash (rounding issues? order of operations?) and the creation of a new automatic cluster which replaces the old one. Since `MatchMarkers` eagerly asserts ownership over the markers, it is happy to take them from the other automatic cluster. The database is updated by adding the new cluster, removing the old cluster, (effectively a nop)... and the cycle repeats. 

So this issue is deeper than initially suspected (though, it still needs the proposed fix to work). In addition, `face.MatchMarkers` needs to abort if exactly the same markers were matched as before. This will be tricky, because `MergeFaces` creates the DB entry *before* trying to match the markers, so even if `MatchMarkers` doesn't call `marker.SetFace` ever, there will be the new Face entry in the DB that will be pruned by `PurgeOrphanFaces`.

### Hypothesis: 

Adding another `ORDER BY` in `faces.ManuallyAddedFaces` (say, by created_at) to ensure a stable order within the `faces` array in the `Optimize` function. For each merge group, that will ensure that the order of `merge.Embeddings()` stays stable, which hopefully will result in a stable result for `m.embedding, m.SampleRadius, m.Samples = face.EmbeddingsMidpoint(embeddings)`, and `m.EmbeddingJSON, err = json.Marshal(m.embedding)` so that the `face_id` of the resulting cluster is hopefully stable. Then, `merged = entity.FirstOrCreateFace(merged)` should return the old entity, instead of a new one, `MatchMarkers` will do nothing (since it already has all those embeddings) and `PurgeOrphanFaces` will correctly return the 0 rows modified. 

### Proposal

change:
```
	err = stmt.Order("subj_uid, samples DESC").Find(&result).Error
```

 to:

```
	err = stmt.Order("subj_uid, samples DESC, created_at ASC").Find(&result).Error
```

Will report back.

theshadow27 committed:
marker update optimizes one face only

theshadow27 commented:
I have tested the hypothesis and, as I suspected after writing it (as explained [here](https://github.com/photoprism/photoprism/issues/3124#issuecomment-2566025503)) it was not sufficient to stop the loop. However, with the change to focus on only the current `subj_uid`, It is getting harder to trigger/reproduce this bug naturally (it was happening all the time before!) I have added over 300 marker labels with no blocking behavior. This is likely because I've tagged all the high-quality photos, and the remaining ones are low quality and being ignored. 

It is still possible to reproduce it by finding an automatically tagged face (e.g. by file hash) and then re-tagging it with the same person. 

When triggered by the background worker, it does seem to be returning much faster. I can measure the time, it must be at least 1/2 of  what it was before. Sometimes the loop will end at 10 iterations (rather than 11). 

I will continue bashing on it... In the meantime, I would truly appreciate a set of experienced eyes. Especially on the new `subj_uid` parameter being passed down to `ManuallyAddedFaces` - I'm suspicious that this query isn't working 100% correctly.

lastzero commented:
@theshadow27 Thanks a lot for submitting these improvements! If you believe this is ready (well tested and stable enough) to be merged, I'd be happy to do a quick review (check for any obvious issues) as soon as possible so we can ship it with an upcoming release (but not the next one, so as not to delay it any further).

lastzero commented:
In the meantime, feel free to add additional unit tests if you have the time to do so and there are edge/special cases that should be (better) covered. This would give additional confidence in the correctness of the code and could help catch unexpected problems before we release it.

lastzero commented:
@theshadow27 I just pushed the vector package that I've been working on, but haven't had time to finish, i.e. use in our existing code. It will be useful to improve the face embedding comparison, although that may be outside the scope of this particular pull request. Unfortunately, it will also require a refactoring of the existing vector code in "internal", which I have started on the feature/pkg-vector branch (there may be smarter ways to do this, it's really just a starting point, as I don't have time to continue working on it right now).

lastzero commented:
Thank you very much for your patience while we completed the frontend upgrade! It's pretty much done now, so I can finally focus on other tasks like merging this PR :)
